[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/IsmailAkram/Assignment_3/blob/main/Assignment_3.ipynb)

In [1]:
# Imports Section
import pandas as pd
import numpy as np

Context reference for future reviews: https://hmcnich.gitbook.io/csc448-f21/assignment-3-equation-of-a-slime

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
slime_df = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")

In [3]:
# Output the first 15 rows of the data
slime_df.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
slime_df.info() # for technical summary
slime_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

In [5]:
slime_df.size # number of datapoints

3000

## Part 2. Splitting the dataset

In [6]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = slime_df[["Temperature °C", "Mols KCL"]] # X (2D array)
label = slime_df["Size nm^3"] # Y

In [7]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.1, train_size=0.9) # splits into 4 arrays

In [8]:
X_train, X_test, y_train, y_test # for display purposes

(     Temperature °C  Mols KCL
 733             329       848
 207             807       529
 289             372       944
 845             144        70
 647             912       965
 ..              ...       ...
 341             473       732
 718             927       360
 605             261        40
 141             423       820
 324             703       774
 
 [900 rows x 2 columns],
      Temperature °C  Mols KCL
 87              846        30
 820             563       395
 446             173       111
 232             101       828
 711             551       694
 ..              ...       ...
 375             657       137
 981             859       811
 677             151       492
 966             152       380
 724             941       892
 
 [100 rows x 2 columns],
 733    5.824778e+05
 207    8.714855e+05
 289    7.322610e+05
 845    2.202800e+04
 647    1.797710e+06
            ...     
 341    7.134573e+05
 718    6.822669e+05
 605    2.405771e+04
 141    7.180

## Part 3. Perform a Linear Regression
https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression

In [9]:
# Use sklearn to train a model on the training set
from sklearn.linear_model import LinearRegression

LinReg = LinearRegression().fit(X_train, y_train) # for fitting; use train.
LinReg.score(X_test, y_test) # for scoring; use test.

0.8594731306915852

In [10]:
# Create a sample datapoint and predict the output of that sample with the trained model
LinReg.predict(np.array([[1,2]])) # nparray to create a 2D array for features (which is a 2D array).

array([-414955.50302426])

In [11]:
LinReg.coef_ # 2 coefficients

array([ 874.04495115, 1033.99759651])

In [12]:
LinReg.intercept_ 

-417897.54316843016

In [13]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

 $LinReg.intercept + LinReg.coef$ 
 
Line of best fit (regression of intercept plus regression of coefficient). It's not 100$ accuracte because it's a 2D matrix, and our model is linear. The equation was created with the coefficents a and b plus the intercept which fits the prediction.

$ y = -416809.6401386601 + 872.67469055a + 1033.21219669b $

$ y = -413870.5410547301 $ ... predict ($h(x)$)

## Part 4. Use Cross Validation
https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation

In [14]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
from sklearn.model_selection import cross_val_score

cross_val_score(LinReg, X_test, y_test, cv=5)

array([0.79339008, 0.88900038, 0.86683782, 0.82558008, 0.85965408])

In [15]:
# Report on their finding and their significance

They're trying to get close to the true value of LinReg from the 5 cv runs.

## Part 5. Using Polynomial Regression
https://scikit-learn.org/stable/modules/linear_model.html#bayesian-regression

In [16]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
from sklearn.preprocessing import PolynomialFeatures
from sklearn import linear_model

PolyReg = PolynomialFeatures(degree=2)
PolyReg_train = PolyReg.fit_transform(X_train)
PolyReg_test = PolyReg.fit_transform(X_test)

BayReg = linear_model.BayesianRidge().fit(PolyReg_train, y_train)
BayReg.score(PolyReg_test, y_test) # 1.0 meaning 100% accuracy

1.0

In [17]:
# Report on the metrics and output the resultant equation as you did in Part 3.
BayReg.coef_

array([ 0.00000000e+00,  1.20000000e+01, -1.46477861e-07, -5.23131885e-12,
        2.00000000e+00,  2.85714287e-02])

In [18]:
BayReg.intercept_

2.1951331291347742e-05

In [19]:
BayReg.predict(np.array([[1, 2, 3, 4, 5, 6]]))

array([34.17145007])

$ y = 2.3343833163380623*10^{-5} + 0.00000000*1 + 1.20000000 * 10^{1}*a + -1.76090506 * 10^{7}*b + -2.64489472 *10^{-11}*a^2 + 2.00000000*ab + 2.85714288 *10^{-2}*b^2 $

Because this is a polynomial regression (using polynomial features), we can convert the 2D matrix into a regular matrix. So we can easily fit the data into a model for a 100% accuracy (as noted by the Bayesian Regression score of 1.0). 

For a degree two polynomial the coefficients arrays refer to these terms: [1, a, b, a^2, ab, b^2]